### PR TITLE
Tree: Use correct cell ID when referring to unattached cells

### PR DESCRIPTION
**Pull request by alex-pardes**
## Description

Added a `detachIdOverride` field to sequence field marks which can represent unattaches (Delete and ReturnFrom). When creating the inverse of an attach this field is set to the attach's cell ID. We use this as the output cell ID of the detach instead of its own ID and revision.

Also created an `adjacentCells` field on `CellId` which lists the IDs of all adjacent cells detached in the same revision. This is used instead of including those adjacent detaches in `CellId.lineage`. This eliminates the need to handle splitting and merging of lineage event when splitting and merging marks.

### CODE REVIEW

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts [hunk 1]
```diff
@@ -96,12 +96,28 @@ export interface HasLineage {
...
+	id: ChangesetLocalIdSchema,
+	count: Type.Number(),
+});
+
 /**
  * @alpha
  */
-export interface CellId extends ChangeAtomId, HasLineage {}
+export interface CellId extends ChangeAtomId, HasLineage {
+	adjacentCells?: IdRange[];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:113]: Consider adding some doc comment to mention the fact that the `CellId`s that make up the range all share the same `IdRange[]` instance.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts [hunk 2]
```diff
@@ -96,12 +96,28 @@ export interface HasLineage {
...
+}
 
-export const CellId = Type.Composite([EncodedChangeAtomId, HasLineage]);
+export const CellId = Type.Composite([
+	EncodedChangeAtomId,
+	HasLineage,
+	Type.Object({ adjacentCells: Type.Optional(IdRange) }),
+]);
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:120]: It just occured to me that we probably should add `noAdditionalProps` to this. No worries if you don't want to do it as part of this PR though.
```suggestion
], noAdditionalProps);
```

File: experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts [hunk 3]
```diff
@@ -317,7 +337,7 @@ export function tryExtendMark<T>(lhs: Mark<T>, rhs: Readonly<Mark<T>>): boolean
 		return false;
 	}
 	const type = rhs.type;
-	if (!areMergeableCellIds(lhs.cellId, lhs.count, rhs.cellId)) {
+	if (!areMergeableChangeAtoms(lhs.cellId, lhs.count, rhs.cellId)) {
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts:340]: CellId still contains lineage, why are we not checking the lineage for equality anymore?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts [hunk 4]
```diff
@@ -186,12 +195,16 @@ function invertMark<TNodeChange>(
 						),
 					];
 				}
+
+				const deleteMark: Delete<TNodeChange> = {
+					type: "Delete",
+					count: mark.count,
+					id: mark.cellId.localId,
+					detachIdOverride: mark.cellId,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts:203]: Should this CellID (when specified) not be used when inverting such a delete? (Line 154 at the time of this comment)
This line (203) is fine, I just can't put comments on line 154.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts [hunk 5]
```diff
@@ -83,17 +83,36 @@ describe("SequenceField - Invert", () => {
...
+				detachIdOverride: { localId: brand(0), revision: tag1 },
+				changes: inverseChildChange1,
+			},
+			{
+				type: "Delete",
+				count: 1,
+				id: brand(1),
+				detachIdOverride: { localId: brand(1), revision: tag1 },
+			},
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:100]:
```suggestion
		const expected: TestChangeset = [
			Mark.delete(1, brand(0), {
				changes: inverseChildChange1,
				detachIdOverride: { localId: brand(0), revision: tag1 },
			}),
			Mark.delete(1, brand(1), { detachIdOverride: { localId: brand(1), revision: tag1 } }),
		];
```

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts [hunk 6]
```diff
@@ -83,17 +83,36 @@ describe("SequenceField - Invert", () => {
...
+		const cellId: ChangeAtomId = { revision: tag1, localId: brand(0) };
+		const input = Change.intentionalRevive(0, 2, cellId);
+		const expected: TestChangeset = [
+			{
+				type: "Delete",
+				id: brand(0),
+				count: 2,
+				detachIdOverride: cellId,
+			},
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:115]:
```suggestion
		const expected: TestChangeset = [Mark.delete(2, brand(0), { detachIdOverride: cellId })];
```

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts [hunk 7]
```diff
@@ -119,14 +138,29 @@ describe("SequenceField - Invert", () => {
...
+				count: 1,
+				changes: inverseChildChange1,
+			},
+			{
+				type: "ReturnFrom",
+				id: brand(1),
+				detachIdOverride: { revision: tag1, localId: brand(1) },
+				count: 1,
+			},
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:163]:
```suggestion
		const expected: TestChangeset = [
			Mark.returnTo(2, brand(0), cellId),
			{ count: 3 },
			Mark.returnFrom(1, brand(0), {
				detachIdOverride: cellId,
				changes: inverseChildChange1,
			}),
			Mark.returnFrom(1, brand(1), {
				detachIdOverride: { revision: tag1, localId: brand(1) },
			}),
		];
```

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts (hunk 1 above)
Comment (alex-pardes) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:113]: That's an implementation detail which we are not currently relying on anywhere. Why does it need to be documented?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts (hunk 4 above)
Comment (alex-pardes) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts:203]: Good point

File: experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts (hunk 1 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/format.ts:113]: I don't think it really *needs* to be, but it's a largely invisible fact that (a) ensures we don't get N² data when someone makes a transaction that deletes a bunch of contiguous nodes, and (b) could lead to subtle bugs if someone were to mutate one without realizing they're mutating all of them (which admittedly seems like an unlikely situation).

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts (hunk 6 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:107]: Not sure what happened with VS-code, removing this an trying again.

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts [hunk 8]
```diff
@@ -83,17 +83,28 @@ describe("SequenceField - Invert", () => {
...
+		const cellId: ChangeAtomId = { revision: tag1, localId: brand(0) };
+		const input = Change.intentionalRevive(0, 2, cellId);
+		const expected: TestChangeset = [
+			{
+				type: "Delete",
+				id: brand(0),
+				count: 2,
+				detachIdOverride: cellId,
+			},
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:107]:
```suggestion
		const expected: TestChangeset = [Mark.delete(2, brand(0), { detachIdOverride: cellId })];
```

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts [hunk 9]
```diff
@@ -119,14 +130,29 @@ describe("SequenceField - Invert", () => {
...
+				count: 1,
+				changes: inverseChildChange1,
+			},
+			{
+				type: "ReturnFrom",
+				id: brand(1),
+				detachIdOverride: { revision: tag1, localId: brand(1) },
+				count: 1,
+			},
+		];
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/invert.spec.ts:155]:
```suggestion
		const expected: TestChangeset = [
			Mark.returnTo(2, brand(0), cellId),
			{ count: 3 },
			Mark.returnFrom(1, brand(0), {
				detachIdOverride: cellId,
				changes: inverseChildChange1,
			}),
			Mark.returnFrom(1, brand(1), {
				detachIdOverride: { revision: tag1, localId: brand(1) },
			}),
		];
```

File: experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts [hunk 10]
```diff
@@ -254,15 +267,20 @@ function invertMark<TNodeChange>(
...
 				);
 			}
+
+			const cellId =
+				mark.type === "ReturnFrom" && mark.detachIdOverride !== undefined
+					? mark.detachIdOverride
+					: {
+							revision: mark.revision ?? revision ?? fail("Revision must be defined"),
+							localId: mark.id,
+					  };
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts:277]: Could use `getDetachCellId` here.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts [hunk 11]
```diff
@@ -887,6 +914,16 @@ export function splitMark<T, TMark extends Mark<T>>(mark: TMark, length: number)
 			if (mark.cellId !== undefined) {
 				(mark2 as Detach).cellId = splitDetachEvent(mark.cellId, length);
 			}
+
+			if (mark2.type === "ReturnFrom") {
+				const mark2Return = mark2 as ReturnFrom<T>;
+				if (mark2Return.detachIdOverride !== undefined) {
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts:920]: `tryExtendMark` also needs to be updated to account for `detachIdOverride`.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts (hunk 4 above)
Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/invert.ts:212]: There is an existing "delete => revive" test. We should add a variant of it that has the override specificed on it.

**Review by yann-achard-MS — APPROVED** [review]

**Pull request merged**